File: src/my_cpp_pkg/src/add_two_ints_server.cpp
#include "rclcpp/rclcpp.hpp"  // Include the RCLCPP library

#include "example_interfaces/srv/add_two_ints.hpp"  // Include the AddTwoInts service message

using std::placeholders::_1;
using std::placeholders::_2;

class AddTwoIntsServerNode : public rclcpp::Node  // Define a class named AddTwoIntsServer that extends rclcpp::Node
{
public:
    AddTwoIntsServerNode() : Node("add_two_ints_server")  // Constructor for AddTwoIntsServer
    {
        server_ = this->create_service<example_interfaces::srv::AddTwoInts>(  // Create a service named "add_two_ints"
            "add_two_ints",
            std::bind(&AddTwoIntsServerNode::callbackAddTwoInts, this, _1, _2));  // Bind the callbackAddTwoInts member function
        RCLCPP_INFO(this->get_logger(), "Service server has been started.");  // Log a message
    }

private:
    void callbackAddTwoInts(const example_interfaces::srv::AddTwoInts::Request::SharedPtr request,
                           const example_interfaces::srv::AddTwoInts::Response::SharedPtr response)
    {
        response->sum = request->a + request->b;  // Calculate the sum of the two integers
        RCLCPP_INFO(this->get_logger(), "%ld + %ld = %ld", request->a, request->b, response->sum);  // Log the result
    }

    rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr server_;  // Declare a service member variable
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);  // Initialize the ROS 2 node
    auto node = std::make_shared<AddTwoIntsServerNode>();  // Create an instance of AddTwoIntsServer
    rclcpp::spin(node);  // Run the node
    rclcpp::shutdown();  // Shutdown the ROS 2 node
    return 0;  // Return 0 to indicate successful execution
}

File: src/my_py_pkg/my_py_pkg/add_two_ints_server.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

# Define a class AddTwoIntsServerNode that inherits from Node class
class AddTwoIntsServerNode(Node):
    def __init__(self):
        super().__init__("add_two_ints_server")  # Initialize the Node with the name "add_two_ints_server"
        # Create a service named "add_two_ints" with the service type AddTwoInts
        # and associate it with the callback function "callback_add_two_ints"
        self.server_ = self.create_service(AddTwoInts, "add_two_ints", self.callback_add_two_ints)
        self.get_logger().info("Add two ints server has been started")
   
    def callback_add_two_ints(self, request, response):
        # Perform the addition operation by adding the values of request.a and request.b
        response.sum = request.a + request.b
        # Log the input values and the result
        self.get_logger().info(str(request.a) + " + " + str(request.b) + " = " + str(response.sum))
        # Return the response object with the calculated sum
        return response


def main(args=None):
    rclpy.init(args=args)
    # Create an instance of the AddTwoIntsServerNode class
    node = AddTwoIntsServerNode()
    # Start spinning the node, allowing it to process incoming requests
    rclpy.spin(node)
    # Clean up and shut down the node
    rclpy.shutdown()


if __name__ == "__main__":
    main()
